rust/lance/src/dataset.rs

```diff
@@ -95,6 +96,9 @@ pub struct Dataset {
     pub(crate) base: Path,
     pub(crate) manifest: Arc<Manifest>,
     pub(crate) session: Arc<Session>,
+
+    /// Cache index metadadta of a version of the dataset.
+    pub(crate) index_metadata_cache: Arc<futures::lock::Mutex<Option<CachedIndexMetadata>>>,
```

Comment:
Why not include this as part of `session`?

Reply:
Good point. Was thinking that this cache has the same lifetime with the `Dataset`. 

But if put into Session, we can actually maintain the cache for multiple opened dataset (i.e., per request). Seems a better deal. I will update it.

Reply:
If you put it in session, you can then just re-use the existing metadata cache.

Reply:
The `session` does have the same lifetime as `Dataset` doesn't it?

Reply:
Session object can be passed in as external managed memory (outside of Dataset)
https://github.com/lancedb/lance/blob/main/rust/lance/src/dataset/builder.rs#L162